pandas/core/ops/__init__.py

```diff
@@ -338,27 +337,28 @@ def column_op(a, b):
             # Note: we do not do this unconditionally as it may be lossy or
             #  expensive for EA dtypes.
             right = np.asarray(right)
-
-            def column_op(a, b):
-                return {i: func(a.iloc[:, i], b[i]) for i in range(len(a.columns))}
-
         else:
+            right = right._values
 
-            def column_op(a, b):
-                return {i: func(a.iloc[:, i], b.iloc[i]) for i in range(len(a.columns))}
+        arrays = []
+        for l, r in zip(left._iter_column_arrays(), right):
+            arrays.append(array_op(l, r))
 
     elif isinstance(right, ABCSeries):
         assert right.index.equals(left.index)  # Handle other cases later
+        right = right._values
 
-        def column_op(a, b):
-            return {i: func(a.iloc[:, i], b) for i in range(len(a.columns))}
+        arrays = []
+        for l in left._iter_column_arrays():
+            arrays.append(array_op(l, right))
 
     else:
         # Remaining cases have less-obvious dispatch rules
         raise NotImplementedError(right)
 
-    new_data = expressions.evaluate(column_op, str_rep, left, right)
```

Comment:
does this have a perf impact?  IIRC a while back I tried removing it because I thought it shouldn't, but found that it did

Reply:
As far as I understand, the `get_array_op` should take care of this? 

For example, the `arrays_ops.py::na_arithmetic_op` also calls `expressions.evaluate` to perform the actual op.

Reply:
That was what i thought too, and I could be remembering incorrectly.  But this is worth double-checking.